### DISCUSSION TITLE
TSDF Voxel Center Offset

### DISCUSSION THREAD

**Discussion by Algomorph** [isl-org/Open3D#3329] [Q&A]
Question: internally, how are voxel centers treated in TSDF?

TLDR; for instance, when a mesh is extracted, it seems like currently there is no extra offset:
https://github.com/intel-isl/Open3D/blob/ad4455f09bf47db99c791257fda96e77859ffb80/cpp/open3d/t/geometry/kernel/TSDFVoxelGridImpl.h#L714-L716
https://github.com/intel-isl/Open3D/blob/ad4455f09bf47db99c791257fda96e77859ffb80/cpp/open3d/t/geometry/kernel/TSDFVoxelGridImpl.h#L761-L763

Am I correct in thinking then that `voxel_x_center_meters` =  `voxel_size` * `voxel_global_x_coordinate`?

**Comment by Algomorph** [comment]
Upon further experimentation & extending Open3D code, I believe voxel centers are directly represented by the multiple of voxel_size and the global coordinates, with no offset. This ultimately means spatial hash block coordinates will have an offset of 1/2 voxel from the nearest voxel center (since these have to include an N^3 cube of whole voxels), which may be relevant to any algorithms involving the hash blocks.

**Comment by theNded** [comment]
It is generally a matter of understanding. My interpretation is that the position at which we store the TSDF value is in practice the voxel center (which is, conceptually, usually regarded as a corner of the voxel where we need to add and subtract 0.5 voxel size back and forth). The surface extraction is based on the interpolation of isosurface values at these positions, and the TSDF values at these positions are computed with projective data association -- that being said, I would prefer to say that there is no bias/offset.

In fact, when I imagine the representation in my head, I usually view it as a collection of grid points, instead of cubes. That perspective might be closer to the implementation.

**Reply by Algomorph** [reply]
This makes sense, and I usually assume the same, and you're right, it's definitely useful to imagine voxels as values at distinct points. The resulting 1/2 voxel size offset from hash block border (or, if you will, 1/2 distance between voxel centers) only comes into play when working with the spatial hash blocks themselves. e.g. when tracing a camera ray to determine which blocks need "activation". Omitting it is a mistake I've seen in another voxel-hashed TSDF implementation (InfiniTAM, which used to be rather popular). I'm not yet sure about Open3D in that regard, but I'll poke around in the allocation code in due time.

**Reply by Algomorph** [reply]
Ah, yes, I see at a glance from the implementation of open3d::t::geometry::kernel::tsdf::Touch function (at least the CPU version) that they omit the 1/2 voxel_size offset. It's a subtle bug that rarely actually manifests itself, blocks so often being much bigger than TSDF truncation distance. But perhaps I'll propose a PR in the future to fix this.

**Reply by theNded** [reply]
I personally still don't think it is a bug -- since we are using `floor`, we always ensure that there are 8 grid points (or voxel corners if you prefer) enclosing the point. This is also true for another application at the voxel scale (https://github.com/intel-isl/Open3D/blob/master/cpp/open3d/t/pipelines/slac/ControlGrid.cpp#L69), so this touch is fine with both voxel and block scales.

I would prefer to say it is an implementation preference or conceptual difference -- when you allocate a block w.r.t a point, do you prefer to place an exact voxel to enclose the point, or just ensure that there are voxels surrounding it. After all, being able to perform a query and trilinear interpolation is the ultimate target.

**Reply by Algomorph** [reply]
Let's see if we can arrive at some common denominator here (pun intended).

I'm looking at this code in the TouchCPU function:
https://github.com/intel-isl/Open3D/blob/f9ef634dcb71b17036633baf1323fdbd344c24e1/cpp/open3d/t/geometry/kernel/TSDFVoxelGridCPU.cpp#L93-L94

Suppose we're dealing with a voxel size of 1 and a block size of 2 (we can imagine this being in `cm` as opposed to `m` like the code if that seems too big). Let me also take everything down to 2D to make things simpler to discuss and understand.
Let's say that `x + sdf_trunc` in the code ends up at the value `1.75`, like in the diagram below.
`static_cast<int>(std::floor(1.75 / 2.0))` will yield value `0` for the x coordinate of the block, whereas the point is actually inside block at `1,0`.

![image](https://user-images.githubusercontent.com/1050606/118191600-d58b0180-b412-11eb-844d-8346b686e891.png)

Am I missing something?

**Reply by theNded** [reply]
Thanks for taking the time for the visualization!

When implementing this snippet, in my mind the blocks are also aligned with the voxels:
![image](https://user-images.githubusercontent.com/6127282/118193497-cd809100-b415-11eb-8d7c-9b1ef5cb1993.png)

And the block lookup in raycasting is consistently using `floor`: https://github.com/intel-isl/Open3D/blob/master/cpp/open3d/t/geometry/kernel/TSDFVoxelGridImpl.h#L1291

So my coordinate system is kind of consistent in itself, but quite confusing if you think in the InfiniTAM convention. Describe this layout in InfiniTAM's language, a voxel block is a bounding box that encloses the voxel centers (instead of their natural corners).

In my initial attempts, I did what you described (e.g. https://github.com/theNded/Open3D/blob/cuda/src/Cuda/Integration/ScalableTSDFVolumeCudaDevice.cuh#L33). But later I realized that it is more like a convention to follow, so I eliminated these by shifting 0.5 voxel size in my mind instead of the code.

Please let me know if it is still unclear / incorrect :)

**Reply by theNded** [reply]
Of course, the boundary distribution in a block will be different too, but they are equivalent in area and consistent upon a shift. I also put it here in case you have questions:
![image](https://user-images.githubusercontent.com/6127282/118196038-51d51300-b41a-11eb-9614-8e20808ac244.png)

**Reply by Algomorph** [reply]
I see, thank you also for taking the time with this.

I think InfiniTAM's convention is actually consistent with yours, unless you're using my fork (correct me if I'm wrong, please).

I thought about it that way at first too, placing voxels at corner points of my spatial hash blocks. But then I discovered (the hard way) that voxels are getting missed. Let's say you get a ray that's passing through that capricious point, (1.75, -0.25) at an angle like so:
![image](https://user-images.githubusercontent.com/1050606/118194973-4ed92300-b418-11eb-851e-985bc63b275f.png)

But because the block containing the voxel (2, 0) was never activated, even though the voxel center with high likelihood projects onto the same pixel from which the green point's ray originates, it won't receive a value during the integration step.

Also, second issue: I don't see how just sampling the eight points around each projected surface point (downsampled too!) gives you accurate block activation, without ray tracing. What if you had a truncation distance that's greater than the block size? Or is that purposefully restricted, so it all works out? If not, you could end up activating blocks at the boundaries of the truncation region, but not in the center.

They do some rudimentary ray tracing (or, rather, ray casting) in InfiniTAM too, but even that tends to miss voxels, because they simply sample points at even intervals along the way. If you're at all interested, I also have a simple algorithmic+code fix for that (computationally, not expensive at all, I believe).

**Reply by theNded** [reply]
> I think InfiniTAM's convention is actually consistent with yours, unless you're using my fork (correct me if I'm wrong, please).

Ok, then my previous words should be referring to your coordinate system rather than InfiniTAM's.

> But because the block containing the voxel (2, 1) was never activated, even though the voxel center with high likelihood projects onto the same pixel from which the green point's ray originates, it won't receive a value during the integration step.

I don't fully get the setup, but I'm pretty sure we can construct an equivalent case in the non-shifted version with a different coordinate -- this is based on the different boundary distribution as I have posted.

> Also, second issue: I don't see how just sampling the eight points around each projected surface point (downsampled too!) gives you accurate block activation, without ray tracing. What if you had a truncation distance that's greater than the block size? Or is that purposefully restricted, so it all works out? If not, you could end up activating blocks at the boundaries of the truncation region, but not in the center.

This is from the legacy Open3D implementation, and in fact, it is more aggressive. If allocate using ray casting, we end up with blue blocks, but now we end up with gray. Experimentally, at least 1.5x blocks are allocated in this setup. I'm switching to ray-casted allocation in the recent development.
![image](https://user-images.githubusercontent.com/6127282/118196627-551cce80-b41b-11eb-90ad-5cfa8cd059d7.png)

> They do some rudimentary ray tracing (or, rather, ray casting) in InfiniTAM too, but even that tends to miss voxels, because they simply sample points at even intervals along the way. If you're at all interested, I also have a simple algorithmic+code fix for that (computationally, not expensive at all, I believe).

I believe it is a trade-off between accuracy and speed. In practice, their ray casting takes only 1ms to run on CUDA. We can of course use AABB intersection in ray casting, but that takes at least 2x time, if not more. Since raycasting is serving tracking and visualization (at least in InfiniTAM), and they are already pretty nice, I believe AABB tracing is not necessary. Would be helpful in an offline rendering system though.

**Reply by Algomorph** [reply]
Actually, I think my solution is much faster than the conventional AABB tracing (it's on-par with the original InfiniTAM solution), but I have not exactly done any performance comparisons, so it's hard to say. 

But here, I don't see how you actually can reconstruct the case using your convention.
We have:
![image](https://user-images.githubusercontent.com/1050606/118197612-4c2cfc80-b41d-11eb-8fc1-390e803f4d1d.png)

The block highlighted in yellow gets missed, so does the voxel (2, 0). You could also use the argument, of course, that not adding the half-voxel offset to every one of the eight coordinates increases performance. This is a very minor issue. These "missed" voxels are pretty rare.

P.S. You can think of it also this way: as it stands, you only consider 1/4 of the volume of the voxel (2,0) when you are allocating blocks. In the 3D case, this translates to 1/4 of the volume of the voxels along the edge and 1/8 for voxels in the (minimum) corner of each block.

**Reply by theNded** [reply]
Sounds great! 
For the coordinates, I still keep my opinion, and you can keep yours too :)
For the rays, I feel like it would be great to have a better ray-based version. In fact, I'm looking forward to a PR if it is comparable with InfiniTAM!

**Reply by Algomorph** [reply]
@theNded, sounds good!

I know some people are against just adding options, but would you be opposed to keeping the current allocation schema that you have as default and simply adding a setting to alter it to mine? We could, perhaps, work out a compromise in design between performance and code simplicity by using a `const` private member of the TSDFVoxelGrid class initialized in the constructor, together with a templated pair of versions for the `open3d::t::geometry::kernel::tsdf::Touch` functions, with choice of function version invoked depending on the `const` private member value (this, not losing performance during the multi-core/multi-threaded traversal of the points in the downsampled point cloud).

I am completely open to integrating a ray-based version as well. In the end, perhaps this could yield a study of whether a collection of such algorithmic / geometric details affects reconstruction performance and for which scenarios (e.g. more complex schema might not produce an impact for certain certain frame-rates or, say, with no objects moving in the scene), and serve as promotion of both your and my schemes for their respective benefits.

**Reply by theNded** [reply]
> We could, perhaps, work out a compromise in design between performance and code simplicity by using a `const` private member of the TSDFVoxelGrid class initialized in the constructor, together with a templated pair of versions for the `open3d::t::geometry::kernel::tsdf::Touch` functions, with choice of function version invoked depending on the `const` private member value (this, not losing performance during the multi-core/multi-threaded traversal of the points in the downsampled point cloud).

If you wish to change, I think we can enclose the offset parameter in the kernel, in this shape:
```cpp
template <bool use_offset>
void TouchCUDA(...) {
    float offset = use_offset ? voxel_size * 0.5 : 0;
   ...
    launcher.LaunchGeneralKernel(n, [=] OPEN3D_DEVICE(int64_t i) {
        ...                        
        int x_low = static_cast<int>(floorf((x - sdf_trunc + offset) / block_size));
    }
}
```
and expose a `bool use_offset` to the user at runtime in the main `TSDFVoxelGrid`. The change should also be applied to `TouchCPU` and `Integrate`. I don't think we will lose any performance at this scale.

> 
> I am completely open to integrating a ray-based version as well. In the end, perhaps this could yield a study of whether a collection of such algorithmic / geometric details affects reconstruction performance and for which scenarios (e.g. more complex schema might not produce an impact for certain certain frame-rates or, say, with no objects moving in the scene), and serve as promotion of both your and my schemes for their respective benefits.

In fact I have one coarse implementation and simple profiling, and I am looking forward to a better one, so any contribution would be great.

**Reply by Algomorph** [reply]
Absolutely, that's exactly what I had in mind. I'll submit a PR.

**Reply by Algomorph** [reply]
@theNded , just wanted to let you know, I haven't forgotten about this issue and it's still on my TODO list.
Currently, I am working on an extension that will allow Open3D to be used in a [DynamicFusion](https://www.cv-foundation.org/openaccess/content_cvpr_2015/papers/Newcombe_DynamicFusion_Reconstruction_and_2015_CVPR_paper.pdf) context.

The sources are in a separate extension right now, but I think can be easily transferred to Open3D in the future when complete.
In case you're interested in taking a look at some of the sources, here are some links:

https://github.com/Algomorph/NeuralTracking/blob/main/csrc/geometry/WarpableTSDFVoxelGrid.cpp
https://github.com/Algomorph/NeuralTracking/blob/main/csrc/geometry/kernel/WarpableTSDFVoxelGridImpl.h
https://github.com/Algomorph/NeuralTracking/blob/main/csrc/geometry/Graph.cpp
https://github.com/Algomorph/NeuralTracking/blob/main/csrc/geometry/kernel/GraphImpl.h
https://github.com/Algomorph/NeuralTracking/blob/main/csrc/geometry/kernel/GraphUtilitiesImpl.h


It's a work-in-progress, I can update you once there is some tangible result.

**Reply by theNded** [reply]
That's super exciting! Please prioritize this over the small fixes.

I also have a [deformable grid](https://github.com/intel-isl/Open3D/tree/master/cpp/open3d/t/pipelines/slac) implementing [the SLAC paper](http://vladlen.info/publications/simultaneous-localization-and-calibration-self-calibration-of-consumer-depth-cameras/). This is not as sophisticated as DynamicFusion, just FYI.

I will be more than happy to help if you plan to merge these functionalities to Open3D!

**Reply by theNded** [reply]
Hey I want to let you know that there are some major upgrades to the hashmap interface and ongoing changes to the TSDF voxel grids. Please feel free to reach out either here or via email if you see any compatibility issues.

**Reply by Algomorph** [reply]
@theNded thanks for letting me know. We've dealt with a few compatibility issues so far, perhaps not so major. I can tell your code is changing for the better, though, so we'll make sure to work those changes in when we're ready. I'm on vacation right now until Thurs next week, but hopeful we'll have something really neat for you by the end of the year, still working out the quirks in dynamic geometry tracking over longer sequences.

**Reply by theNded** [reply]
Thanks for the update! I'm rewriting the voxel grid part as well for differentiable volumetric rendering, will let you know when it's in alpha shape.

**Comment by germanros1987** [comment]
Should we close this discussion @Algomorph ?

**Reply by Algomorph** [reply]
@germanros1987, the issue isn't really resolved because maintainers failed to address or understand the original problem. I think[ this post ](https://github.com/isl-org/Open3D/discussions/3329#discussioncomment-736093) demonstrates the issue best. My recommendation would be to first understand the problem -- it's not a "convention" problem, it's a real geometry problem, in my view -- then close this convo and open a new issue, referencing all of the necessary details.

The dynamic scene reconstruction discussion is a separate topic and we can open a separate discussion for it later, when I am closer to being able to merge my fork into Open3D.

**Reply by Algomorph** [reply]
To really drive the point home, I estimate that your current schema actually causes about 23% incorrect block allocations given 8x8x8 hash block size. This percentage scales down given bigger hash block size. It's about 11.5% missing/unallocated blocks and about 11.5% unneeded/incorrect blocks being allocated if the block is aligned to the camera axis, and these ratios change (but their sum remains approximately the same) for other block orientations.

The reason you don't immediately see the effects is that, during subsequent allocations, often the correct blocks are getting hit.

The fix is super-simple, i.e. the cost of 3 additions per check, to account for the voxel centers not lying at the corner of the voxel hash blocks.
